src/languages/fsharp.js

```diff
@@ -221,16 +225,130 @@ export default function(hljs) {
     relevance: 0
   };
 
+  const makeTypeAnnotationMode = function(prefix, prefixScope) {
+    return {
+      begin: regex.concat( // a type annotation is a
+        prefix,            // should be a colon or the 'of' keyword
+        regex.lookahead(   // that has to be followed by
+          regex.concat(
+            /\s*/,         // optional space
+            regex.either(  // then either of:
+              /\w/,        // word
+              /'/,         // generic type name
+              /\^/,        // generic type name
+              /#/,         // flexible type name
+              /``/,        // quoted type name
+              /\(/,        // parens type expression
+              /{\|/,       // anonymous type annotation
+      )))),
+      beginScope: prefixScope,
+      // BUG: because ending with \n is necessary for some cases, multi-line type annotations are not properly supported.
+      // Examples where \n is required at the end:
+      // - abstract member definitions in classes: abstract Property : int * string
+      // - return type annotations: let f f' = f' () : returnTypeAnnotation
+      // - record fields definitions: { A : int \n B : string }
+      end: regex.lookahead(
+        regex.either(
+          /\n/,
+          /=/)),
+      relevance: 0,
+      // we need the known types, and we need the type constraint keywords and literals. e.g.: when 'a : null
+      keywords: hljs.inherit(ALL_KEYWORDS, { type: KNOWN_TYPES }),
+      contains: [
+        COMMENT,
+        GENERIC_TYPE_SYMBOL,
+        hljs.inherit(QUOTED_IDENTIFIER, { scope: null }), // match to avoid strange patterns inside that may break the parsing
+        OPERATOR
+      ]
+    };
+  }
+
+  const TYPE_ANNOTATION = makeTypeAnnotationMode(/:/, 'operator');
+  const DISCRIMINATED_UNION_TYPE_ANNOTATION = makeTypeAnnotationMode(/\bof\b/, 'keyword');
+
+  // type MyType<'a> = ...
+  const TYPE_DECLARATION = {
+    begin: [
+      /(^|\s+)/, // prevents matching the following: `match s.stype with`
+      /type/,
+      /\s+/,
+      IDENTIFIER_RE
+    ],
+    beginScope: {
+      2: 'keyword',
+      4: 'title.class'
+    },
+    end: regex.lookahead(/\(|=|$/),
+    keywords: ALL_KEYWORDS, // match keywords in type constraints. e.g.: when 'a : null
+    contains: [
+      COMMENT,
+      hljs.inherit(QUOTED_IDENTIFIER, { scope: null }), // match to avoid strange patterns inside that may break the parsing
+      GENERIC_TYPE_SYMBOL,
+      {
+        // For visual consistency, highlight type brackets as operators.
+        scope: 'operator',
+        match: /<|>/
+      },
+      TYPE_ANNOTATION // generic types can have constraints, which are type annotations. e.g. type MyType<'T when 'T : delegate<obj * string>> =
+    ]
+  };
+
   const COMPUTATION_EXPRESSION = {
     // computation expressions:
     scope: 'computation-expression',
+    // BUG: might conflict with record deconstruction. e.g. let f { Name = name } = name // will highlight f
```

Comment:
Can we just capture this edge case with a separate rule to handle `let [ident] {` ?

Reply:
We certainly could, but it's only one edge case among others.
Examples:
```fsharp
type Record = { Prop: int }
// f2 is a function with 3 parameters (a, b, c). a and c are both extracted from records
let private f2 { Prop = a } b { Prop = c } = () // `private` can be used to the function private...
let value = 42
let x = f2 { Prop = 42 } value { Prop = 0 } // Call f2 with some values
```
With the current grammar it will highlight as follows: 
![image](https://user-images.githubusercontent.com/1535823/142045305-4f807316-de8d-4298-8596-b058ac8b1d31.png)
(Ideally, only the words `type`, `int`, `let` and `private` should be in red)

I've seen other projects use a list of known computation expressions to avoid this problem, but IMO it's not very good either: anyone can create a new one and there is no consensus on a list of "official" computer expressions...

Even knowing all these shortcomings, I still feel like it's preferable to continue using the current regex to highlight `identifier {`.
I don't think it's warranted to make an exception just for `let ident {`.

What do you think?

Reply:
> Even knowing all these shortcomings, I still feel like it's preferable ...

But why?  

I see you refer to these as "edge cases", is that strictly true?

Reply:
> I see you refer to these as "edge cases", is that strictly true?

It's an edge case in the sense that computation expressions are probably more frequent than something followed by a record (which wrongly triggers the computation expression highlighting).

I think you started referring to this bug as an edge case in your first comment :p

> But why?

Because I think it's important to highlight computation expressions, even if it means there are false positives, and because I think it's better to have failures that are consistent (at least the user can try to figure out why it breaks and use a different syntax).
But all things considered — and since it appears we can't really disambiguate between a computation expression and something followed by a record — it might be better to drop support for computation expression highlighting altogether... Would that be preferable?

(Note the problem was already present before the current PR)

Reply:
> (Note the problem was already present before the current PR)

Since you point this out we'll leave it for now and see how it goes... feels like it's right on the border though.